src/windows/top_toolbar.c

```diff
@@ -626,70 +628,801 @@ static void window_top_toolbar_paint()
+
+			// If SHIFT pressed
+			if (RCT2_GLOBAL(RCT2_ADDRESS_SCENERY_TOOL_SHIFT_PRESSED, uint8) != 0){
+				z += RCT2_GLOBAL(RCT2_ADDRESS_SHIFT_PRESS_Z_VECTOR, sint16);
+			}
+
+			if (z < 16){
+				z = 16;
+			}
+
+			RCT2_GLOBAL(RCT2_ADDRESS_SCENERY_Z_COORDINATE, sint16) = z;
+		}
+
+		if (*grid_x == 0x8000)
+			return;
+
+		RCT2_GLOBAL(0x00F64F15, uint8) = window_scenery_secondary_colour;
+		RCT2_GLOBAL(0x00F64F16, uint8) = window_scenery_tertiary_colour;
+		// Also places it in lower but think thats for clobering
+		*parameter_1 = (selected_scenery & 0xFF) << 8;
+		*parameter_2 = cl | (window_scenery_primary_colour << 8);
+		*parameter_3 = 0;
+		break;
+	}
+	case 3:
+	{
+		// Large scenery
+
+		// If CTRL not pressed
+		if (RCT2_GLOBAL(RCT2_ADDRESS_SCENERY_TOOL_CTRL_PRESSED, uint8) == 0){
+			sub_68A15E(x, y, grid_x, grid_y, NULL, NULL);
+
+			if (*grid_x == 0x8000)
+				return;
+
+			RCT2_GLOBAL(RCT2_ADDRESS_SCENERY_Z_COORDINATE, sint16) = 0;
+
+			// If SHIFT pressed
+			if (RCT2_GLOBAL(RCT2_ADDRESS_SCENERY_TOOL_SHIFT_PRESSED, uint8) != 0){
+				rct_map_element* map_element = map_get_surface_element_at(*grid_x / 32, *grid_y / 32);
+
+				if (map_element == NULL){
+					*grid_x = 0x8000;
+					return;
+				}
+
+				sint16 z = (map_element->base_height * 8) & 0xFFF0;
+				z += RCT2_GLOBAL(RCT2_ADDRESS_SHIFT_PRESS_Z_VECTOR, sint16);
+
+				if (z < 16){
+					z = 16;
+				}
+
+				RCT2_GLOBAL(RCT2_ADDRESS_SCENERY_Z_COORDINATE, sint16) = z;
+			}
+		}
+		else{
+			sint16 z = RCT2_GLOBAL(RCT2_ADDRESS_CTRL_PRESS_Z_COORDINATE, sint16);
+			sub_6894D4(x, y, z, grid_x, grid_y);
+
+			// If SHIFT pressed
+			if (RCT2_GLOBAL(RCT2_ADDRESS_SCENERY_TOOL_SHIFT_PRESSED, uint8) != 0){
+				z += RCT2_GLOBAL(RCT2_ADDRESS_SHIFT_PRESS_Z_VECTOR, sint16);
+			}
+
+			if (z < 16){
+				z = 16;
+			}
+
+			RCT2_GLOBAL(RCT2_ADDRESS_SCENERY_Z_COORDINATE, sint16) = z;
+		}
+
+		if (*grid_x == 0x8000)
+			return;
+
+		*grid_x &= 0xFFE0;
+		*grid_y &= 0xFFE0;
+
+		uint8 rotation = window_scenery_rotation;
+		rotation -= RCT2_GLOBAL(RCT2_ADDRESS_CURRENT_ROTATION, uint8);
+		rotation &= 0x3;
+
+		*parameter_1 = (rotation << 8);
+		*parameter_2 = window_scenery_primary_colour | (window_scenery_secondary_colour << 8);
+		*parameter_3 = selected_scenery & 0xFF;
+		break;
+	}
+	case 4:
+	{
+		// Banner
+
+		uint16 flags = 0xFF9F;
+		int interaction_type = 0;
+		rct_map_element* map_element;
+
+		int map_x, map_y;
+		get_map_coordinates_from_pos(x, y, flags, &map_x, &map_y, &interaction_type, &map_element, NULL);
+		*grid_x = (sint16)map_x;
+		*grid_y = (sint16)map_y;
+
+		if (interaction_type == VIEWPORT_INTERACTION_ITEM_NONE)
+		{
+			*grid_x = 0x8000;
+			return;
+		}
+
+		uint8 rotation = window_scenery_rotation;
+		rotation -= RCT2_GLOBAL(RCT2_ADDRESS_CURRENT_ROTATION, uint8);
+		rotation &= 0x3;
+		
+		sint16 z = map_element->base_height;
+
+		if (map_element->properties.path.type & (1 << 2)){
+			if (rotation != ((map_element->properties.path.type & 3) ^ 2)){
```

Comment:
I'm probably not the first to have figured this out, but: the third bit indicates whether the path is sloped and the first two bits contain the direction when walking upwards. Xor-ing a direction with 2 seems to yield the opposite direction.